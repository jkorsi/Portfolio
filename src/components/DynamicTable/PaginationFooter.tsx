import React from "react";
import firstIcon from "../../icons/pagination-icons/arrow-end-left-icon.svg";
import lastIcon from "../../icons/pagination-icons/arrow-end-right-icon.svg";
import previousIcon from "../../icons/pagination-icons/line-angle-left-icon.svg";
import nextIcon from "../../icons/pagination-icons/line-angle-right-icon.svg";

interface PaginationProps {
  currentPage: number;
  totalPages: number;
  itemsPerPage: number;
  handlePageChange: (page: number) => void;
  handleItemsPerPageChange: (value: number) => void;
}

const PaginationFooter: React.FC<PaginationProps> = ({
  currentPage,
  totalPages,
  itemsPerPage,
  handlePageChange,
  handleItemsPerPageChange,
}) => {
  const pageRange = 2; // Number of pages to show before and after the current page
  const lastPage = totalPages - 1;

  const handlePreviousPage = () => {
    if (currentPage > 0) {
      handlePageChange(currentPage - 1);
    }
  };

  const handleNextPage = () => {
    if (currentPage < lastPage) {
      handlePageChange(currentPage + 1);
    }
  };

  const handleFirstPage = () => {
    if (currentPage !== 0) {
      handlePageChange(0);
    }
  };

  const handleLastPage = () => {
    if (currentPage !== lastPage) {
      handlePageChange(lastPage);
    }
  };

  const handlePageSelect = (page: number) => {
    if (page !== currentPage) {
      handlePageChange(page);
    }
  };

  const handleItemCalcsForEachPageChange = (
    e: React.ChangeEvent<HTMLSelectElement>
  ) => {
    const value = parseInt(e.target.value);
    console.log(value);
    handleItemsPerPageChange(value);
  };

  const generatePageNumbers = () => {
    const pageNumbers = [];
    for (let i = currentPage - pageRange; i <= currentPage + pageRange; i++) {
      if (i >= 0 && i < totalPages) {
        pageNumbers.push(i);
      }
    }
    return pageNumbers;
  };

  return (
    <div className="flex items-center justify-between px-4 py-2">
      <div className="flex items-center space-x-2">
        <button className="p-0 rounded w-7 h-7" onClick={handleFirstPage}>
          <img
            src={firstIcon}
            alt="First Page"
            className="px-2 py-1 rounded  hover:bg-gray-200 inline-block w-7 h-7"
          />
        </button>
        <button className="p-0 rounded w-7 h-7" onClick={handlePreviousPage}>
          <img
            src={previousIcon}
            alt="Previous Page"
            className="px-2 py-1 rounded  hover:bg-gray-200 inline-block w-7 h-6"
          />
        </button>
        {generatePageNumbers().map((page) => (
          <button
            key={page}
            className={`px-2 py-1 rounded border  ${
              currentPage === page ? "bg-gray-200" : "hover:bg-gray-200"
            }`}
            onClick={() => handlePageSelect(page)}
          >
            {page}
          </button>
        ))}
        <button className="p-0 rounded w-7 h-7" onClick={handleNextPage}>
          <img
            src={nextIcon}
            alt="Next Page"
            className="px-2 py-1 rounded hover:bg-gray-200 inline-block w-7 h-6"
          />
        </button>
        <button onClick={handleLastPage} className="p-0 rounded w-7 h-7">
          <img
            src={lastIcon}
            alt="Last Page"
            className="px-2 py-1 rounded hover:bg-gray-200 inline-block w-7 h-7"
          />
        </button>
      </div>
      <div>
        <span>Items per Page:</span>
        <select
          value={itemsPerPage}
          onChange={handleItemCalcsForEachPageChange}
          className="px-2 py-1 ml-2 border rounded"
        >
          <option value={5}>5</option>
          <option value={10}>10</option>
          <option value={25}>25</option>
          <option value={50}>50</option>
        </select>
      </div>
    </div>
  );
};

export default PaginationFooter;
