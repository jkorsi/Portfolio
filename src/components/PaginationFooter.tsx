import React from "react";

interface PaginationProps {
  currentPage: number;
  totalPages: number;
  itemsPerPage: number;
  handlePageChange: (page: number) => void;
  handleItemsPerPageChange: (value: number) => void;
}

const PaginationFooter: React.FC<PaginationProps> = ({
  currentPage,
  totalPages,
  itemsPerPage,
  handlePageChange,
  handleItemsPerPageChange,
}) => {
  const pageRange = 2; // Number of pages to show before and after the current page
  const lastPage = totalPages - 1;

  const handlePreviousPage = () => {
    if (currentPage > 0) {
      handlePageChange(currentPage - 1);
    }
  };

  const handleNextPage = () => {
    if (currentPage < lastPage) {
      handlePageChange(currentPage + 1);
    }
  };

  const handleFirstPage = () => {
    if (currentPage !== 0) {
      handlePageChange(0);
    }
  };

  const handleLastPage = () => {
    if (currentPage !== lastPage) {
      handlePageChange(lastPage);
    }
  };

  const handlePageSelect = (page: number) => {
    if (page !== currentPage) {
      handlePageChange(page);
    }
  };

  const handleItemCalcsForEachPageChange = (
    e: React.ChangeEvent<HTMLSelectElement>
  ) => {
    const value = parseInt(e.target.value);
    console.log(value);
    handleItemsPerPageChange(value);
  };

  const generatePageNumbers = () => {
    const pageNumbers = [];
    for (let i = currentPage - pageRange; i <= currentPage + pageRange; i++) {
      if (i >= 0 && i < totalPages) {
        pageNumbers.push(i);
      }
    }
    return pageNumbers;
  };

  return (
    <div className="flex items-center justify-between px-4 py-2">
      <div className="flex items-center space-x-2">
        <button
          className="px-2 py-1 rounded border hover:bg-gray-200"
          onClick={handleFirstPage}
        >
          First
        </button>
        <button
          className="px-2 py-1 rounded border hover:bg-gray-200"
          onClick={handlePreviousPage}
        >
          Previous
        </button>
        {generatePageNumbers().map((page) => (
          <button
            key={page}
            className={`px-2 py-1 rounded border ${
              currentPage === page ? "bg-gray-200" : "hover:bg-gray-200"
            }`}
            onClick={() => handlePageSelect(page)}
          >
            {page}
          </button>
        ))}
        <button
          className="px-2 py-1 rounded border hover:bg-gray-200"
          onClick={handleNextPage}
        >
          Next
        </button>
        <button
          className="px-2 py-1 rounded border hover:bg-gray-200"
          onClick={handleLastPage}
        >
          Last
        </button>
      </div>
      <div>
        <span>Items per Page:</span>
        <select
          value={itemsPerPage}
          onChange={handleItemCalcsForEachPageChange}
          className="px-2 py-1 ml-2 border rounded"
        >
          <option value={3}>3</option>
          <option value={10}>10</option>
          <option value={25}>25</option>
          <option value={50}>50</option>
        </select>
      </div>
    </div>
  );
};

export default PaginationFooter;
