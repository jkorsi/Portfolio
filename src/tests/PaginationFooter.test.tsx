import { render, fireEvent } from "@testing-library/react";
import PaginationFooter from "../components/PaginationFooter";

describe("PaginationFooter", () => {
  const mockHandlePageChange = jest.fn();
  const mockHandleItemsPerPageChange = jest.fn();

  const mockProps = {
    currentPage: 2,
    totalPages: 5,
    itemsPerPage: 10,
    handlePageChange: mockHandlePageChange,
    handleItemsPerPageChange: mockHandleItemsPerPageChange,
  };

  it("renders correctly", () => {
    const { asFragment } = render(<PaginationFooter {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls handlePageChange with the correct value when previous button is clicked", () => {
    const { getByText } = render(<PaginationFooter {...mockProps} />);
    const previousButton = getByText("Previous");
    fireEvent.click(previousButton);
    expect(mockHandlePageChange).toHaveBeenCalledWith(1);
  });

  it("calls handlePageChange with the correct value when next button is clicked", () => {
    const { getByText } = render(<PaginationFooter {...mockProps} />);
    const nextButton = getByText("Next");
    fireEvent.click(nextButton);
    expect(mockHandlePageChange).toHaveBeenCalledWith(3);
  });

  it("calls handlePageChange with the correct value when first button is clicked", () => {
    const { getByText } = render(<PaginationFooter {...mockProps} />);
    const firstButton = getByText("First");
    fireEvent.click(firstButton);
    expect(mockHandlePageChange).toHaveBeenCalledWith(0);
  });

  it("calls handlePageChange with the correct value when last button is clicked", () => {
    const { getByText } = render(<PaginationFooter {...mockProps} />);
    const lastButton = getByText("Last");
    fireEvent.click(lastButton);
    expect(mockHandlePageChange).toHaveBeenCalledWith(4);
  });

  it("calls handlePageChange with the correct value when a page number is clicked", () => {
    const { getByText } = render(<PaginationFooter {...mockProps} />);
    const pageNumberButton = getByText("1");
    fireEvent.click(pageNumberButton);
    expect(mockHandlePageChange).toHaveBeenCalledWith(1);
  });

  it("calls handleItemsPerPageChange with the correct value when items per page is changed", () => {
    const { getByRole } = render(<PaginationFooter {...mockProps} />);
    const select = getByRole("combobox");
    fireEvent.change(select, { target: { value: "3" } });
    expect(mockHandleItemsPerPageChange).toHaveBeenCalledWith(3);
  });
});
