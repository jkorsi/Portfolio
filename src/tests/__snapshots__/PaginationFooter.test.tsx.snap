// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationFooter renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-between px-4 py-2"
  >
    <div
      class="flex items-center space-x-2"
    >
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        First
      </button>
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        Previous
      </button>
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        0
      </button>
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        1
      </button>
      <button
        class="px-2 py-1 rounded border bg-gray-200"
      >
        2
      </button>
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        3
      </button>
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        4
      </button>
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        Next
      </button>
      <button
        class="px-2 py-1 rounded border hover:bg-gray-200"
      >
        Last
      </button>
    </div>
    <div>
      <span>
        Items per Page:
      </span>
      <select
        class="px-2 py-1 ml-2 border rounded"
      >
        <option
          value="3"
        >
          3
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
        <option
          value="50"
        >
          50
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
